hashes/src/chacha20poly1305/chacha20.rs

```diff
@@ -0,0 +1,380 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! The ChaCha20 stream cipher based on RFC7539.
+
+/// The first four words (32-bit) of the ChaCha stream cipher state are constants.
+const WORD_1: u32 = 0x61707865;
+const WORD_2: u32 = 0x3320646e;
+const WORD_3: u32 = 0x79622d32;
+const WORD_4: u32 = 0x6b206574;
+
+/// Each quarter round of ChaCha scrambles 4 words (32-bit) of the state
+/// using some Addition (mod 2^32), Rotation, and XOR (a.k.a. ARX). 8 quarter
+/// rounds make up a round. A block is broken up into 16 32-bit words
+/// and each quarter round takes 4 words as input.
+const CHACHA_ROUND_INDICIES: [(usize, usize, usize, usize); 8] = [
+    // The first 4 rounds are rows of a 4x4 matrix
+    // of the block broken up into 32-bit words.
+    (0, 4, 8, 12),
+    (1, 5, 9, 13),
+    (2, 6, 10, 14),
+    (3, 7, 11, 15),
+    // The next 4 rounds are diagonals of a 4x4 matrix
+    // of the block broken up into 32-bit words.
+    (0, 5, 10, 15),
+    (1, 6, 11, 12),
+    (2, 7, 8, 13),
+    (3, 4, 9, 14),
+];
+
+/// The cipher's block size is 64 bytes.
+const CHACHA_BLOCKSIZE: usize = 64;
+
+/// The ChaCha20 stream cipher.
```

Comment:
I haven't seen the different variants of chacha20 have different names, unfortunately.

Though I *believe* that "the one in Bitcoin" has been consistent since we first started using it in secp256k1-zkp for batch validation.

Reply:
If there's no better name than "the one in Bitcoin" then I'm dropping my request to use a generic name for the crate and request to have this information somewhere near the top of the documentation ("This crate implements the same variant of Chacha20-Poly1305 that is used in Bitcoin").

Reply:
OK, then that name, please.

Reply:
I updated the docs to be clear this is the 20-round, IETF (RFC8439) variant with the 96-bit nonce and 32-bit block counter. I believe this is a compete description? Some more history here: https://github.com/bitcoin/bitcoin/issues/19225